### ISSUE TITLE
MS-DOS 6.22 crash problems when using a 119MB hard disk drive (build 5226/above)

### ISSUE THREAD

**Issue by win98user** [86Box/86Box#3875]
### What happened?

Hello there, and this is my 4th 86Box bug issue.

So I'm having a problem in 86Box (build 5226/above), when I was using MS-DOS 6.22 on a 119MB hard-disk drive, MS-DOS 6.22 crashes like a Windows 1.0 blue screen of death, when it passed and finished the (HIMEM is testing extended memory... done.), not the normal (C:\>) screen.
(NOTE: I tested that problem using some of these machines: AMI 386DX clone, Commodore PC 30 III, Phoenix IBM AT, and Siemens PCD-2L).

### Configuration file

```ini
[General]
vid_renderer = qt_software
dpi_scale = 0
sound_gain = 18
confirm_exit = 0

[Machine]
machine = acc386
cpu_family = i386dx
cpu_speed = 40000000
cpu_multi = 1
cpu_use_dynarec = 0
fpu_softfloat = 1
time_sync = local
fpu_type = 387
pit_mode = 1
cpu_waitstates = 2
cpu = 4
mem_size = 8192

[Video]
gfxcard = rtg3106

[Input devices]
mouse_type = msserial

[Sound]
fm_driver = nuked
midi_device = cm32ln
mpu401_standalone = 1

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = xtide_at
cassette_mode = load

[Realtek RTG3106 (ISA)]
memory = 512

[Microsoft Serial Mouse]
port = 0
buttons = 3

[Roland CM-32LN Emulation]
output_gain = 100
reverb = 1
reverb_output_gain = 100
reversed_stereo = 1
nice_ramp = 1

[Ports (COM & LPT)]
lpt1_device = text_prt

[Hard disks]
hdd_01_parameters = 17, 14, 1024, 0, ide
hdd_01_fn = hard-disk/ms-dos_6.22.vhd
hdd_01_vhd_blocksize = 4096
hdd_01_speed = ramdisk
hdd_01_ide_channel = 0:0

[Floppy and CD-ROM drives]
fdd_01_type = 35_2hd
fdd_02_type = 525_2hd
fdd_01_image_history_01 = C:/Games/win98user 2/floppy-disk/Microsoft MS-DOS 6.22 Plus Enhanced Tools (3.5)/Disk2.img
fdd_01_image_history_02 = C:/Games/win98user 2/floppy-disk/Microsoft MS-DOS 6.22 Plus Enhanced Tools (3.5)/Disk1.img
fdd_01_image_history_03 = floppy-disk/Microsoft Windows 3.1 (3.10.103) (3.5)/Disk01.img
fdd_01_turbo = 1
fdd_01_fn = C:/Games/win98user 2/floppy-disk/Microsoft MS-DOS 6.22 Plus Enhanced Tools (3.5)/Disk3.img

[Other peripherals]
postcard_enabled = 1
```

### Operating system

Windows 11

### CPU

AMD 5600H

### 86Box version

Build 5226

### Build architecture

Windows - x64 (64-bit)

### Build type

- [ ] New recompiler
- [ ] Debug build

### Download source

Official website (Jenkins, GitHub)

### Additional context

CFG/NVR/HDD/Screenshots ZIP file:
[files.zip](https://github.com/86Box/86Box/files/13627022/files.zip)

This is a result of running MS-DOS 6.22 on a 119MB hard-disk drive:

AMI 386DX clone:
![MS-DOS 6 22 (with drive name (MS-DOS_6))](https://github.com/86Box/86Box/assets/152977043/ca305b75-3fe5-45cb-879d-915a5a93554c)
![MS-DOS 6 22 (without drive name (MS-DOS_6))](https://github.com/86Box/86Box/assets/152977043/6e1284e6-d194-4188-853b-cde31e9306bc)

Commodore PC 30 III:
![MS-DOS 6 22 in Commodore PC 30 III (4096KB)](https://github.com/86Box/86Box/assets/152977043/25f34ddb-a51a-4c7f-9f05-d012fdbfdafc)
![MS-DOS 6 22 in Commodore PC 30 III (8192KB)](https://github.com/86Box/86Box/assets/152977043/431b4f2a-7e66-41e9-8bc5-9ea836769694)

Phoenix IBM AT:
![MS-DOS 6 22 in Phoenix IBM AT (4096KB)](https://github.com/86Box/86Box/assets/152977043/1443f7c8-846e-4efe-9c4b-be954401fc01)
![MS-DOS 6 22 in Phoenix IBM AT (8192KB)](https://github.com/86Box/86Box/assets/152977043/12a9c93b-1936-4e3b-9055-22bd9409fb21)

Siemens PCD-2L:
![MS-DOS 6 22 in Siemens PCD-2L (4096KB)](https://github.com/86Box/86Box/assets/152977043/f93aaab3-5b97-47cb-b38f-1a86d71729f4)
![MS-DOS 6 22 in Siemens PCD-2L (8192KB)](https://github.com/86Box/86Box/assets/152977043/fe1aa809-6a8a-4733-a4f0-1b56791a320f)

**Comment by Fenix770** [comment]
The Phoenix 286 AT works fine with XTIDE (using the AT variant), with HIMEM.SYS. On the setup BIOS you must put "none" on the hard disk 1 and 2 options, since the XTIDE has its own HDD BIOS.

**Comment by Fenix770** [comment]
Not, yet. 
Windows 1.0x won't work on DOS 4+. As workaround, on DOS 5+, you should add this file on SETVER utility:
Type: "SETVER WIN100.BIN 3.40", and restart the VM. Then Win 1.0x will works.
![MS-DOS 6X+WIN 10](https://github.com/86Box/86Box/assets/31600563/f24c450a-f293-4ba8-b050-4ffcc1d638e7)

**Comment by Fenix770** [comment]
The Siemens machine seems to have issues at boot time.

**Comment by Fenix770** [comment]
I found that the Siemens have issues with certain graphics cards. But works fine with the ISA VGA.
 By the way I'm using the x32 version on 86box. I think this bug maybe related to 64 bits version.

**Comment by Fenix770** [comment]
That is strange. Did you try to delete the NVR files?

**Comment by Fenix770** [comment]
![Siemens config](https://github.com/86Box/86Box/assets/31600563/04be0ad4-c104-4de4-8b39-d5602df398f6)

**Comment by Fenix770** [comment]
![IBM AT Phoenix 1](https://github.com/86Box/86Box/assets/31600563/c1001ad5-983c-4d30-bffe-6cd51072d9b2)

![IBM AT Phoenix 2](https://github.com/86Box/86Box/assets/31600563/3ed03dbe-ba59-4564-9684-c706335e2351)



Also, at which clock speed (Mhz) did you configure these machines. The IBM AT supports only 6/8 Mhz. The Siemens was launched around 1986, so it would have a 6-10Mhz limit. Only late 286 launched after 1988 can reach 16/20 Mhz.

**Comment by Fenix770** [comment]
I don't know what may cause this!. My host machine is a i5 4210U with 12GB DDR 3 1600 Mhz.
At this point, I can only guess, 86box does not emulate the RAM speeds of the original machines. (For example, a 486 SX/16 maybe as fast as a real 486 SX/33.)
So, I don't know which configuration do you have, but maybe if you have a modern machine with DDR 5 RAM. The HIMEM.SYS driver may crash, I'm only guessing.

**Comment by OBattler** [comment]
I suspect it's XTIDE causing the problem. Clearly, whatever is being loaded after HIMEM.SYS, conflicts with it.

**Comment by OBattler** [comment]
You also can't just put in a hard disk image formatted on another machine and expect it to magically work.

**Comment by Fenix770** [comment]
But works fine on my end. Also, the HDD in this case is managed by the XTIDE, not the BIOS of the machine.

**Comment by OBattler** [comment]
@win98user Try creating an .IMG image instead of .VHD.

**Comment by OBattler** [comment]
Try a different combination of heads and sectors.

**Comment by HanHeld** [comment]
Me too?

I've been trying to use a type 46 hard drive with different 486 models, but no dice; they seem to all hang, and I will have other odd behavior such as freezing when I try to run either dos 6.22's edit or scandisk (not chkdsk -it runs chkdsk fine). I deleted a machine and then tried doing it all over from scratch but ran into the same problem with different 486 boards and mother boards (I tried socket 1, socket 3 and so on).

However, when I made a larger hd image; 995 cyl, 16 heads, 63 sectors per track ...I was able to install and boot from it just fine (after telling the bios it was there and restoring a backup onto it).

**Comment by HanHeld** [comment]
> > Me too?
> > I've been trying to use a type 46 hard drive (1224 cyl, 15 heads, 17 sectors per track) with different 486 models, but no dice; they seem to all hang, and I will have other odd behavior such as freezing when I try to run either dos 6.22's edit or scandisk (not chkdsk -it runs chkdsk fine). I deleted a machine and then tried doing it all over from scratch but ran into the same problem with different 486 boards and mother boards (I tried socket 1, socket 3 and so on).
> > However, when I made a larger hd image; 995 cyl, 16 heads, 63 sectors per track ...I was able to install and boot from it just fine (after telling the bios it was there and restoring a backup onto it).
> 
> Okay, does it hang on the (HIMEM is testing extended memory... done.) screen too? I'm gonna try that 119MB (1024 C, 14 H, 17 S) problem on the Intel YM430TX motherboard to see if it has the problem.

I was able to hit F8 and step through config.sys -including himem.  But then it would hang when it tried to run autoexec.bat.

**Comment by HanHeld** [comment]
It's morning here, I'll set up a VM with those specs (Intel YM430TX machine, ide drive with 1024c,14 heads, 17spt) and report back after I do a dos 6.22 install.

**Comment by HanHeld** [comment]
OK -I can't find a "YM430TX", but I was able to reproduce your results:
 
![Screenshot 2023-12-19 101221](https://github.com/86Box/86Box/assets/3117821/b7dcff14-a24b-45d2-ba0c-cac956d115c3)
![Screenshot
![Monitor_1_20231219-101524-993](https://github.com/86Box/86Box/assets/3117821/5a465b4a-4199-4a93-aba2-478fbdd57cb6)
 2023-12-19 101241](https://github.com/86Box/86Box/assets/3117821/671f6333-c9ce-4da1-946b-dc7ee6bc4e34)

**Comment by HanHeld** [comment]
![Monitor_1_20231219-101524-993](https://github.com/86Box/86Box/assets/3117821/74547caa-99db-4eee-a863-ac9cf04ffb1d)

**Comment by HanHeld** [comment]
That's on build 5253, win x64, old dynarec

**Comment by HanHeld** [comment]
Ok, Aquamarine pointed me in the right direction so I tested an new Intel YM430TX VM with those hard disk settings:
![Monitor_1_20231219-103903-783](https://github.com/86Box/86Box/assets/3117821/61251fb3-e0a2-465f-afb7-10e0b1358b53)
...and got the same results after installing dos 6.22:
![Monitor_1_20231219-103553-691](https://github.com/86Box/86Box/assets/3117821/f6f0809b-f2fc-4c81-b5b3-f838b5d91357)

**Comment by HanHeld** [comment]
So, I've been testing and I've been able to get it boot with different sized hard drives. I was thinking that maybe it was freaking out about being 1024 cylinders or more -but nope. I am able to make and install dos from hard drive images that are 1024/16/63 and 1224/16/63 as well. 

No dice with 1024/14/17 and no dice with 1224/14/17 -I'm not sure what to try next.

**Comment by Fenix770** [comment]
I found a similar bug!. I created a VM with a type 9 HDD (C/H/S 900/15/17), on several 386 SX and the Leopard IBM 486SLC, the machine is unable to boot. I installed MS-DOS 5.0 using 3.5 720kb floppy disks (dated as November 1991) and the deleted and created a new partition with MS-DOS using 5.25 360kb floppy disks (dated as april 1991)

**Comment by Fenix770** [comment]
I found a similar bug with HIMEM.SYS at boot time. I tested on Flytech 386SX with PC-DOS 5.02 and C/H/S 1020/15/17.
(I know older machines may have issues with more than 1023 cylinders)

![Monitor_1_20231220-201424-980](https://github.com/86Box/86Box/assets/31600563/57cf903b-2462-4141-a288-298715a39f60)

[ATI EGA Wonder 800+]
memory = 256
monitor_type = 9

[General]
vid_renderer = qt_software
vid_resize = 1

[Machine]
machine = flytech386
cpu_family = i386sx
cpu_speed = 25000000
cpu_multi = 1
cpu_use_dynarec = 0
fpu_softfloat = 0
time_sync = local
mem_size = 4096

[Video]
gfxcard = internal

[Input devices]
mouse_type = msserial

[Sound]
fm_driver = nuked

[Network]
net_01_link = 0
net_02_link = 0
net_03_link = 0
net_04_link = 0

[Storage controllers]
hdc = internal
cassette_mode = load

[Hard disks]
hdd_01_parameters = 17, 15, 1020, 0, ide
hdd_01_fn =
hdd_01_speed = 1994_4500rpm
hdd_01_ide_channel = 0:0
hdd_01_vhd_blocksize = 4096


[nvr.zip](https://github.com/86Box/86Box/files/13733418/nvr.zip)

**Comment by Fenix770** [comment]
I suspect the bug is related to the number of heads not cylinders!. The type 35 HDD C/H/S 1024/9/17 (76 Mb) works just fine, but the type 9 HDD C/H/S 900/15/17 (112 Mb) have random issues. For example, Norton commander 2.01/3.0 freezes when loaded from HDD but works just fine if loaded from the floppy disk. 

I  installed MS-DOS 3.31 and tested two 286 SCAT 82c235 chipsets.

Yesterday I created the type 35 HDD and worked fine, I thought that bug was caused by the number of HDD files on the folder I have been using!

**Comment by Fenix770** [comment]
I also found strange issue. On the 76mb HDD image, the game Starmines II works fine when loaded from it. But on the 1024/12/17, 102 Mb HDD the game throws a memory error, but works fine when loaded from the floppy disk.


![Monitor_1_20231223-074208-273](https://github.com/86Box/86Box/assets/31600563/306efbc3-4d73-4981-84d2-6ba9619f06de)

Here is the floppy disk with the game: 
[Starmines II.zip](https://github.com/86Box/86Box/files/13758816/Starmines.II.zip)

**Comment by Fenix770** [comment]
Update:

I created a new VM with the Flytech 386 SX with a HDD with C/H/S 1024/9/26  . This machine has a rather newer BIOS that supports LBA. For some reason, the autodetect function of this BIOS configures the disk as LBA. Works fine as LBA, but cannot boot if configured as NORMAL.

Only later BIOS launched in 1994 or 1995 have LBA support. 86box can read correctly HDDs that I created a year ago, or so. 
I think the bug is caused by the way 86box calculate the HDD size at the creation time.

**Comment by Fenix770** [comment]
Hello!

I've been able to create a type 9 HDD (900/15/17) on build 4311. I know it is very old build at this point!
![Type 9 HDD 86box 311](https://github.com/86Box/86Box/assets/31600563/26287bfe-433c-4614-85f3-2008575c8498)

**Comment by Fenix770** [comment]
I installed Windows 3.1 on this virtual HDD on build 4311. Then I returned to the build 5290, and boots fine.
(I'm not sure if the bug is in the way 86box created the nvr, because it is the exactly the same config, I haven't changed anything else)

**Comment by Fenix770** [comment]
Seems to works fine also on build 5000, win x32

**Comment by Fenix770** [comment]
I haven't tested it yet, but it is not only a HIMEM.SYS bug but rather (in DOS 5.0) the VM is unable to boot and cannot even read the CONFIG.SYS file.

**Comment by Fenix770** [comment]
Yes, once the image file is created with an older build of 86box (build 5000 in this case) I can then change to the latest (build 5290) and boots just fine. I the upgraded to MS-DOS 6.22 with the Upgrade version. MS-DOS 6.22 Scandisk didn't find any error.

![HDD build 5000](https://github.com/86Box/86Box/assets/31600563/dd6e5002-b6d1-496c-8f84-8926fb3f22f8)

![HDD build 5290](https://github.com/86Box/86Box/assets/31600563/db847337-c6e9-4f4a-b5e4-e9e47c7a13ac)

Here is the HDD image file test it with the NVR file, maybe the reason of the bug.
[nvr.zip](https://github.com/86Box/86Box/files/13770837/nvr.zip)
[0-Experiment type 9 hdd b500.zip](https://github.com/86Box/86Box/files/13770838/0-Experiment.type.9.hdd.b500.zip)

And also the cfg file:
[86box-cfg.zip](https://github.com/86Box/86Box/files/13770848/86box-cfg.zip)

**Comment by Fenix770** [comment]
I found a strange behaviour on Quadtel 286 with Hercules monochrome card and a type 9 HDD. 

1. I installed the Norton Utilities 4.5. The NU.EXE program chrashes the VM with random characters on the screen
2. The Hoyle 2 game keeps reading the HDD forever, after loading the graphical mouse pointer, but the game works fine if loaded from the 720kb floppy.
3. Norton Disk Doctor does not find any error, and also the Norton Speed Disk can defrag the HDD correctly.

Here is the HDD , NVR and CFG files to test:

HDD:
[286+DOS 4.00 +WIN 286.VHD.zip](https://github.com/86Box/86Box/files/13842021/286%2BDOS.4.00.%2BWIN.286.VHD.zip)

NVR:
[nvr-Quadtel-286.zip](https://github.com/86Box/86Box/files/13842023/nvr-Quadtel-286.zip)

CFG:
[86box.zip](https://github.com/86Box/86Box/files/13842025/86box.zip)
